File: src/chapters/linear-systems/content/code/back-substitution.jl
function back_substitution(U, b)
    n = length(b)
    x = zeros(n)
    for i in n:-1:1
        x[i] = (b[i] - sum(U[i, j] * x[j] for j in i+1:n; init = 0.0)) / U[i, i]
    end
    return x
end

U = [2.0 1 -1; 0 1 2; 0 0 3]; b = [1.0, 8, 9]
println(back_substitution(U, b))   # [1, 2, 3]

File: src/pages/practice/guides/matlab/back-substitution.m
function x = back_substitution(U, b)
% BACK_SUBSTITUTION  Solve an upper-triangular system U x = b.
    n = numel(b); x = zeros(n, 1);
    for i = n:-1:1
        x(i) = (b(i) - U(i, i+1:n) * x(i+1:n)) / U(i, i);
    end
end

% --- Demo ---
U = [2 1 -1; 0 1 2; 0 0 3]; b = [1; 8; 9];
disp(back_substitution(U, b)');    % 1 2 3
